### PR TITLE
Update callback scheme to the one we extensively discussed

### PR DESCRIPTION
**Pull request by ColdenCullen**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/client_channel_handler.c [hunk 1]
```diff
@@ -50,9 +50,16 @@ static int s_packet_handler_connack(
...
+    if (connack.connect_return_code == AWS_MQTT_CONNECT_ACCEPTED) {
+        connection->state = AWS_MQTT_CLIENT_STATE_CONNECTED;
+    }
 
-    MQTT_CLIENT_CALL_CALLBACK(connection, on_connack, connack.connect_return_code, connack.session_present);
+    if (connection->state == AWS_MQTT_CLIENT_STATE_RECONNECTING) {
+        MQTT_CLIENT_CALL_CALLBACK_ARGS(connection, on_resumed, connack.connect_return_code, connack.session_present);
+    } else {
+        MQTT_CLIENT_CALL_CALLBACK_ARGS(
+            connection, on_connection_complete, AWS_OP_SUCCESS, connack.connect_return_code, connack.session_present);
```
Comment (graebm) [inline source/client_channel_handler.c]: If someone's called disconnect() from the RECONNECTING state, that will put us into DISCONNECTING state and the on_connection_complete() callback will fire a 2nd time.

File: include/aws/mqtt/client.h [hunk 2]
```diff
@@ -224,7 +255,10 @@ int aws_mqtt_client_connection_connect(
  *              otherwise AWS_OP_ERR and aws_last_error() is set.
  */
 AWS_MQTT_API
-int aws_mqtt_client_connection_disconnect(struct aws_mqtt_client_connection *connection);
+int aws_mqtt_client_connection_disconnect(
+    struct aws_mqtt_client_connection *connection,
+    aws_mqtt_client_on_disconnect_fn *on_disconnect,
+    void *userdata);
```
Comment (graebm) [inline include/aws/mqtt/client.h]: We probably want disconnect to cancel any pending commands. Currently, that happens during destroy, but in python-land that's not going to happen until the connection is garbage-collected, which means basically never.

File: include/aws/mqtt/client.h [hunk 3]
```diff
@@ -85,7 +95,10 @@ typedef void(aws_mqtt_client_publish_received_fn)(
     struct aws_mqtt_client_connection *connection,
     const struct aws_byte_cursor *topic,
     const struct aws_byte_cursor *payload,
-    void *user_data);
+    void *userdata);
+
+/* Called when a connection is closed, right before any resources are deleted */
+typedef void(aws_mqtt_client_on_disconnect_fn)(struct aws_mqtt_client_connection *connection, void *userdata);
```
[the remark this answers is not in the document]
Comment (ColdenCullen) [reply include/aws/mqtt/client.h]: Note for future colden: the second comment boils down to "we should call on_disconnect even if we've already disconnected."

File: include/aws/mqtt/private/client_impl.h [hunk 4]
```diff
@@ -30,18 +30,25 @@
...
         }                                                                                                              \
     } while (false)
 
 enum aws_mqtt_client_connection_state {
-    AWS_MQTT_CLIENT_STATE_INIT,
     AWS_MQTT_CLIENT_STATE_CONNECTING,
     AWS_MQTT_CLIENT_STATE_CONNECTED,
+    AWS_MQTT_CLIENT_STATE_RECONNECTING,
     AWS_MQTT_CLIENT_STATE_DISCONNECTING,
+    AWS_MQTT_CLIENT_STATE_DISCONNECTED,
```
[the remark this answers is not in the document]
Comment (ColdenCullen) [reply include/aws/mqtt/private/client_impl.h]: Destroy should, yes. The others not necessarily, as they could be set for change on reconnect. Once we have a logger these could definitely be warnings though.

File: source/client.c [hunk 5]
```diff
@@ -197,9 +203,34 @@ static void s_mqtt_client_shutdown(
     /* Alert the connection we've shutdown */
     bool attempt_reconnect =
         connection->state != AWS_MQTT_CLIENT_STATE_DISCONNECTING && error_code != AWS_IO_TLS_ERROR_NEGOTIATION_FAILURE;
-    if (connection->callbacks.on_disconnect) {
-        attempt_reconnect =
-            connection->callbacks.on_disconnect(connection, error_code, connection->callbacks.user_data);
+
```
[the remark this answers is not in the document]
Comment (ColdenCullen) [reply source/client.c]: Good catch, will fix.

File: source/client.c [hunk 6]
```diff
@@ -507,9 +559,16 @@ int aws_mqtt_client_connection_connect(
  * Disconnect
  ******************************************************************************/
 
-int aws_mqtt_client_connection_disconnect(struct aws_mqtt_client_connection *connection) {
+int aws_mqtt_client_connection_disconnect(
+    struct aws_mqtt_client_connection *connection,
+    aws_mqtt_client_on_disconnect_fn *on_disconnect,
+    void *userdata) {
+
+    if (connection->state < AWS_MQTT_CLIENT_STATE_DISCONNECTING) {
```
[the remark this answers is not in the document]
Comment (ColdenCullen) [reply source/client.c]: Nah, I'm gonna call the function RIGHT NOW

File: include/aws/mqtt/mqtt.h [hunk 7]
```diff
@@ -52,6 +52,8 @@ enum aws_mqtt_error {
     AWS_ERROR_MQTT_INVALID_TOPIC,
     AWS_ERROR_MQTT_TIMEOUT,
     AWS_ERROR_MQTT_PROTOCOL_ERROR,
+    AWS_ERROR_MQTT_NOT_CONNECTED,
+    AWS_ERROR_MQTT_ALREADY_CONNETED,
```
Comment (justinboswell) [inline include/aws/mqtt/mqtt.h]: ALREADY_CONNE**C**TED

File: source/mqtt.c [hunk 8]
```diff
@@ -137,6 +137,9 @@ void aws_mqtt_load_error_strings() {
             AWS_DEFINE_ERROR_INFO_MQTT(
                 AWS_ERROR_MQTT_PROTOCOL_ERROR,
                 "Protocol error occured."),
+            AWS_DEFINE_ERROR_INFO_MQTT(
+                AWS_ERROR_MQTT_ALREADY_CONNETED,
```
Comment (justinboswell) [inline source/mqtt.c]: Should be fixed by fixing the enum

File: include/aws/mqtt/mqtt.h (hunk 7 above)
Comment (ColdenCullen) [reply include/aws/mqtt/mqtt.h]: Boooo I suck

**Pull request merged**